### PR TITLE
[Fix] Weight initializer stddev calculation

### PR DESCRIPTION
**Pull request by zhoonit**
- [Fix] Weight initializer stddev calculation

```
The weight initializer was fitted to only fully connected layer case,
this patch extends weight initializer to properly work with other
weights as well.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/tensor/weight.cpp [hunk 1]
```diff
@@ -40,6 +42,29 @@ void Weight::runVariableInitializer() {
...
+  /// effective dimension then actual numbers here.
+  switch (dim.rank()) {
+  case 1:
+    fan_out = fan_in = dim.width();
+    break;
+  case 2: /// fully connected layers
+    fan_in = dim.height();
+    fan_out = dim.width();
+    break;
+  case 4: { /// convolution filters, @todo extend this to > 4
```
Comment (kparichay) [inline nntrainer/tensor/weight.cpp:57]: if the dimension of some axis is 1, the rank will be 3 leading to fan_out/fan_in to be 1.

File: nntrainer/tensor/weight.cpp [hunk 2]
```diff
@@ -40,6 +42,29 @@ void Weight::runVariableInitializer() {
...
 
+  unsigned int fan_in, fan_out;
+
+  /// @fixme: when unit is equal to one, this does not work we need to rely on
+  /// effective dimension then actual numbers here.
+  switch (dim.rank()) {
+  case 1:
+    fan_out = fan_in = dim.width();
+    break;
+  case 2: /// fully connected layers
```
Comment (kparichay) [inline nntrainer/tensor/weight.cpp:53]: if rank == 2, but weight and height are 1 which is true for 1x1 conv filter, IMO this case doesn't handle it correctly.

Comment (zhoonit) [reply nntrainer/tensor/weight.cpp:53]: Yes, there is `fixme` right above. Do you have a suggestion instead of relying on effective dimension?

Comment (kparichay) [reply nntrainer/tensor/weight.cpp:53]: aah, I missed that fixme.

Comment (kparichay) [reply nntrainer/tensor/weight.cpp:53]: How about below?
Case 1: B=1, C=1, and rank = 1 - current case 1
Case 2: B=1, C=1, and rank = 2 - current case 2
Default : current case 3

Comment (zhoonit) [reply nntrainer/tensor/weight.cpp]: Updated :)

File: nntrainer/tensor/weight.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/tensor/weight.cpp]: Discussed together with below :)

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/tensor/weight.cpp [hunk 3]
```diff
@@ -40,6 +42,24 @@ void Weight::runVariableInitializer() {
   Tensor &var_ref = getVariableRef();
   const TensorDim dim = var_ref.getDim();
 
+  unsigned int fan_in, fan_out;
+
+  /// @fixme: when unit is equal to one, this does not work, we need to rely on
+  /// effective dimension then actual numbers here. For now, some heuristics
+  /// added to infer what would be fan_in/fan_out
+  if (dim.batch() * dim.channel() * dim.height() == 1) {
+    fan_out = fan_in = dim.width();
```
Comment (jijoongmoon) [inline nntrainer/tensor/weight.cpp:51]: do we need this?? if batch == channel == height ==1, then field_size = dim.width() . then fan_in and fan_out = dim.width().

**Review by jijoongmoon — APPROVED** [review]
LGTM with recommandations.

File: nntrainer/tensor/weight.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/tensor/weight.cpp:51]: Right :) I'll update.

**Pull request merged**